ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A reference simulator of the RealPBFT protocol under \"ideal
+-- circumstances\"
+--
+module Test.Dynamic.Ref.RealPBFT (
+  Outcome (..),
+  State (..),
+  advanceUpTo,
+  emptyState,
+  nextSlot,
+  nullState,
+  step,
+  viable,
+  ) where
+
+import           Data.Foldable (Foldable, foldl', toList)
+import qualified Data.Map as Map
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+
+import           Ouroboros.Network.Block (SlotNo (..))
+
+import           Ouroboros.Consensus.NodeId (CoreNodeId (..))
+import           Ouroboros.Consensus.Protocol.Abstract (SecurityParam (..))
+import           Ouroboros.Consensus.Protocol.PBFT (PBftParams (..))
+
+import           Test.Dynamic.Util.NodeJoinPlan
+
+oneK :: Num a => PBftParams -> a
+oneK PBftParams{pbftSecurityParam} =
+    fromIntegral (maxRollbacks pbftSecurityParam)
+
+twoK :: Num a => PBftParams -> a
+twoK PBftParams{pbftSecurityParam} =
+    2 * fromIntegral (maxRollbacks pbftSecurityParam)
+
+oneN :: Num a => PBftParams -> a
+oneN PBftParams{pbftNumNodes} = fromIntegral pbftNumNodes
+
+{-------------------------------------------------------------------------------
+  PBFT state
+-------------------------------------------------------------------------------}
+
+-- | Outcome of a node leading a slot
+--
+data Outcome
+
+  = Absent
+    -- ^ the leader hadn't yet joined
+
+  | Nominal
```

Comment:
Why `Nominal`?

Reply:
https://en.wiktionary.org/wiki/nominal#Adjective

> 10. (engineering) According to plan or design; normal.
> 
>     We'll just do a nominal flight check.
>     Apart from the slightly high temperature, all the readings from the spacecraft are nominal.

Hmm. Maybe that usage is less common than I thought. I wanted the constructor name to evoke "the boring case where the node forged a valid block that the rest of the net promptly adopted".

Reply:
Rocket scientists use "nominal". Rocket science is cool. Therefore...  :grin: 

We just want to a void a RUD in our node.

https://en.wikipedia.org/wiki/Rapid_unscheduled_disassembly

Reply:
> https://en.wiktionary.org/wiki/nominal#Adjective
> 
> > 1. (engineering) According to plan or design; normal.
> >    We'll just do a nominal flight check.
> >    Apart from the slightly high temperature, all the readings from the spacecraft are nominal.
> 
> Hmm. Maybe that usage is less common than I thought. I wanted the constructor name to evoke "the boring case where the node forged a valid block that the rest of the net promptly adopted".

Sounds totally correct, but as a non-native speaker I was not familiar with that meaning of the word. If you include a tiny comment, e.g., something like "according to plan", it's fine for me.

Reply:
I added a comment.